### PR TITLE
Suci profile A

### PR DESCRIPTION
**Pull request by Stg22**
Here is some source code based on Compact25519 library for x25519 functions and on my KDF X9.63 development.

The modification runs with Open5GS version 2.6.1 which supports SUCI :o)

### CODE REVIEW

File: src/lib/CMakeLists.txt [hunk 1]
```diff
@@ -5,6 +5,8 @@ add_library(common-lib ${HDR_FILES} ${SRC_FILES})
 
 target_compile_options(common-lib PRIVATE -Wall -Wextra -pedantic -Wno-unused-parameter)
 
+target_include_directories(common-lib PUBLIC ${CMAKE_CURRENT_SOURCE_DIR})
```
Comment (aligungr) [inline src/lib/CMakeLists.txt:8]: Could you explain this modification?

File: src/ue.cpp [hunk 2]
```diff
@@ -150,6 +150,12 @@ static nr::ue::UeConfig *ReadConfigYaml()
 
     if (yaml::HasField(config, "supi"))
         result->supi = Supi::Parse(yaml::GetString(config, "supi"));
+    if (yaml::HasField(config, "protectionScheme"))
+        result->protectionScheme = yaml::GetInt32(config, "protectionScheme", 0, 255);
+    if (yaml::HasField(config, "homeNetworkPublicKeyId"))
+        result->homeNetworkPublicKeyId = yaml::GetInt32(config, "homeNetworkPublicKeyId", 0, 255);
+if (yaml::HasField(config, "homeNetworkPublicKey"))
```
Comment (aligungr) [inline src/ue.cpp:157]: Please fix the indentation in line 157.

File: config/free5gc-ue.yaml [hunk 3]
```diff
@@ -4,6 +4,8 @@ supi: 'imsi-208930000000003'
 mcc: '208'
 # Mobile Network Code value of HPLMN (2 or 3 digits)
 mnc: '93'
+# SUCI Protection Scheme : 0 for Null-scheme, 1 for Profile A and 2 for Profile B
+protectionScheme: 0
```
Comment (aligungr) [inline config/free5gc-ue.yaml:8]: homeNetworkPublicKey and homeNetworkPublicKeyId fields are missing. Is that intentional?

**Review by aligungr — CHANGES_REQUESTED** [review]
Thanks for the PR. I have some minor questions. Please see the review comments.

File: config/free5gc-ue.yaml (hunk 3 above)
Comment (Stg22) [reply config/free5gc-ue.yaml:8]: Free5GC does not support SUCI with Profile A or B. Only Null-Scheme is possible so no key is needed.
May be, in order to be more clear, an empty key could be preferable with a comment explaining why the key is not set.

File: src/ue.cpp (hunk 2 above)
Comment (Stg22) [reply src/ue.cpp:157]: It will be fixed. Sorry for this basic mistake :o(

File: src/lib/CMakeLists.txt (hunk 1 above)
Comment (Stg22) [reply src/lib/CMakeLists.txt:8]: It was done during a troubleshooting action but not deleted after. Sorry, it is just another mistake.

File: config/free5gc-ue.yaml (hunk 3 above)
Comment (Stg22) [reply config/free5gc-ue.yaml:8]: My bad!
Free5GC implements SUCI. I will modify the config file.

**Comment by aligungr** [comment]
Thanks for your work, I'll merge it. I have one last question, and wondering your opinon.

I think we should use Null-scheme by default (`protectionScheme: 0`) in sample config files. Because most users are likely to use config files as is without many modifications.

What do you think? Do most users need a protected scheme, or the null scheme?

**Comment by Stg22** [comment]
You are probably right, it is easier to use Null-Scheme by default. Generally with 5G open source projects, you need addtional actions for configuring SUCI. That's why in all configuration files, I have put the parameter protectionScheme to 0.
For a use with many commercial 5G platforms, SUCI is now the default rule (may be because Authorities requires such behaviours in many countries).

**Comment by aligungr** [comment]
Thanks @Stg22

**Comment by FatimaaIsmail27** [comment]
@Stg22 hello, I want to try encryption of SUPI using profile A. what should I change in configuration files ? because I tried protection scheme 1 instead of 0 and it is not working.

**Comment by FatimaaIsmail27** [comment]
@Stg22 yes I checked the configuration file and all the parts you added already exist in my files, also there exists a home network public key:
supi: 'imsi-901700000000001'
# Mobile Country Code value of HPLMN
mcc: '901'
# Mobile Network Code value of HPLMN (2 or 3 digits)
mnc: '70'
# SUCI Protection Scheme : 0 for Null-scheme, 1 for Profile A and 2 for Profile B
protectionScheme: 0
# Home Network Public Key for protecting with SUCI Profile A
homeNetworkPublicKey: '5a8d38864820197c3394b92613b20b91633cbd897119273bf8e4a6f4eec0a650'
# Home Network Public Key ID for protecting with SUCI Profile A
homeNetworkPublicKeyId: 1
# Routing Indicator
routingIndicator: '0000'
That's why I didn't understand what should I change other than the protection scheme id..
And should I change it everywhere the protection scheme id is mentioned in the files?
Sorry for taking from your time. 
Thank you.

**Pull request merged**